panic: runtime error: negative shift amount

**Describe the bug**
While running the init command with master it panics

**To Reproduce**
Steps to reproduce the behavior:
1. with the following definition:
```
type Type string

const (
	TypePerson   Type = "person"
	TypeBusiness Type = "business"
)

type customerIn struct {
	ClientId string `json:"client_id"`
	Type     string `json:"type"`
}

type customerOut struct {
	Id        string    `json:"id"`
	ClientId  string    `json:"client_id"`
	Type      string    `json:"type"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type customersOut struct {
	Result []customerOut
}
```
run: 
```
swag init -g service/server/server.go --pd
swag fmt -g service/server/server.go
```
Throws the error:
```
swag init -g service/server/server.go --pd
2022/11/22 08:09:45 Generate swagger docs....
2022/11/22 08:09:45 Generate general API Info, search dir:./
2022/11/22 08:09:45 warning: failed to get package name in dir: ./, error: execute go list command, exit status 1, stdout:, stderr:no Go files in /Users/nicolas.sassi/dev/soul/acaiah
panic: runtime error: negative shift amount

goroutine 1 [running]:
github.com/swaggo/swag.(*PackageDefinitions).evaluateConstValue(0xc00607e870, 0xc0068aa2d0?, 0xc00665ef60?, {0x165bee0?, 0xc005cecf90?}, {0x165a460, 0xc0001b1800}, 0x1c?)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/package.go:146 +0x102b
github.com/swaggo/swag.(*PackagesDefinitions).EvaluateConstValue(0xc00479f698?, 0xc00607e870, 0xc00682de00, 0x0)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/packages.go:282 +0x19d
github.com/swaggo/swag.(*PackagesDefinitions).evaluateAllConstVariables(...)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/packages.go:265
github.com/swaggo/swag.(*PackagesDefinitions).ParseTypes(0xc0001b1800)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/packages.go:110 +0x254
github.com/swaggo/swag.(*Parser).ParseAPIMultiSearchDir(0xc00027c000, {0xc0002780e0?, 0x1?, 0x0?}, {0x7ff7bfeff555?, 0x18?}, 0x64)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/parser.go:375 +0x3bf
github.com/swaggo/swag/gen.(*Gen).Build(0xc000249950, 0xc0002502a0)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/gen/gen.go:182 +0x637
main.initAction(0xc000266780?)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/cmd/swag/main.go:158 +0x7bd
github.com/urfave/cli/v2.(*Command).Run(0xc000220ea0, 0xc000262300)
        /Users/nicolas.sassi/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/command.go:163 +0x5dc
github.com/urfave/cli/v2.(*App).RunContext(0xc000195d40, {0x165cef0?, 0xc000196008}, {0xc00019c000, 0x5, 0x5})
        /Users/nicolas.sassi/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:313 +0xb7d
github.com/urfave/cli/v2.(*App).Run(...)
        /Users/nicolas.sassi/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224
main.main()
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/cmd/swag/main.go:229 +0x5c5
make: *** [swagger] Error 2
```


**Expected behavior**
Swagger created

**Your swag version**
go install github.com/swaggo/swag/cmd/swag@e5d507dd472777bec3a3744f7f9cc86065037530

**Your go version**
1.19.1

**Desktop (please complete the following information):**
 - OS: macOS Monterey 12.6.1

If I change the models for the following:
```
type Type string

const (
	TypePerson   Type = "person"   // person
	TypeBusiness Type = "business" /* business */
)

type customerIn struct {
	ClientId string `json:"client_id"`
	Type     Type   `json:"type"`
}

type customerOut struct {
	Id        string    `json:"id"`
	ClientId  string    `json:"client_id"`
	Type      Type      `json:"type"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}
```
With the intention of using the new enums feature the error changes for the following:

```
swag init -g service/server/server.go --pd
2022/11/22 08:16:29 Generate swagger docs....
2022/11/22 08:16:29 Generate general API Info, search dir:./
2022/11/22 08:16:29 warning: failed to get package name in dir: ./, error: execute go list command, exit status 1, stdout:, stderr:no Go files in /Users/nicolas.sassi/dev/soul/acaiah
panic: interface conversion: interface {} is string, not int

goroutine 1 [running]:
github.com/swaggo/swag.(*PackageDefinitions).evaluateConstValue(0xc001020000, 0x1a63878?, 0x90?, {0x165bee0?, 0xc001536a20?}, {0x165a460, 0xc000089820}, 0x1066467?)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/package.go:132 +0x113f
github.com/swaggo/swag.(*PackageDefinitions).evaluateConstValue(0xc001020000, 0xc0090a8510?, 0xc006ba3e00?, {0x165bee0?, 0xc001536c60?}, {0x165a460, 0xc000089820}, 0xc0001d1aa0?)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/package.go:115 +0xd9
github.com/swaggo/swag.(*PackagesDefinitions).EvaluateConstValue(0xc005711698?, 0xc001020000, 0xc0058b8bc0, 0x0)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/packages.go:282 +0x19d
github.com/swaggo/swag.(*PackagesDefinitions).evaluateAllConstVariables(...)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/packages.go:265
github.com/swaggo/swag.(*PackagesDefinitions).ParseTypes(0xc000089820)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/packages.go:110 +0x254
github.com/swaggo/swag.(*Parser).ParseAPIMultiSearchDir(0xc000200000, {0xc0001fc0f0?, 0x1?, 0x0?}, {0x7ff7bfeff555?, 0x18?}, 0x64)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/parser.go:375 +0x3bf
github.com/swaggo/swag/gen.(*Gen).Build(0xc0001d1950, 0xc0001d82a0)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/gen/gen.go:182 +0x637
main.initAction(0xc0001eaa00?)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/cmd/swag/main.go:158 +0x7bd
github.com/urfave/cli/v2.(*Command).Run(0xc0001a6a20, 0xc0001e8300)
        /Users/nicolas.sassi/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/command.go:163 +0x5dc
github.com/urfave/cli/v2.(*App).RunContext(0xc0000bd380, {0x165cef0?, 0xc00002c108}, {0xc0000240a0, 0x5, 0x5})
        /Users/nicolas.sassi/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:313 +0xb7d
github.com/urfave/cli/v2.(*App).Run(...)
        /Users/nicolas.sassi/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224
main.main()
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/cmd/swag/main.go:229 +0x5c5
make: *** [swagger] Error 2
```

> there are some warning `failed to parse const...` in the std output

The std output seems pretty much the same
```
swag init -g service/server/server.go  --requiredByDefault --pd
2022/11/22 14:05:53 Generate swagger docs....
2022/11/22 14:05:53 Generate general API Info, search dir:./
2022/11/22 14:05:53 warning: failed to get package name in dir: ./, error: execute go list command, exit status 1, stdout:, stderr:no Go files in /Users/nicolas.sassi/dev/soul/acaiah
panic: interface conversion: interface {} is string, not int

goroutine 1 [running]:
github.com/swaggo/swag.(*PackageDefinitions).evaluateConstValue(0xc000ce3f40, 0x1a65498?, 0x90?, {0x165bee0?, 0xc0014da9c0?}, {0x165a460, 0xc0001b1800}, 0xc00f941588?)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/package.go:132 +0x113f
github.com/swaggo/swag.(*PackageDefinitions).evaluateConstValue(0xc000ce3f40, 0xc01d29f260?, 0xc0068b7b30?, {0x165bee0?, 0xc0014da9f0?}, {0x165a460, 0xc0001b1800}, 0xc000249a70?)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/package.go:115 +0xd9
github.com/swaggo/swag.(*PackagesDefinitions).EvaluateConstValue(0xc00f941698?, 0xc000ce3f40, 0xc00580e740, 0x0)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/packages.go:282 +0x19d
github.com/swaggo/swag.(*PackagesDefinitions).evaluateAllConstVariables(...)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/packages.go:265
github.com/swaggo/swag.(*PackagesDefinitions).ParseTypes(0xc0001b1800)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/packages.go:110 +0x254
github.com/swaggo/swag.(*Parser).ParseAPIMultiSearchDir(0xc00027a000, {0xc0002760d0?, 0x1?, 0x0?}, {0x7ff7bfeff53d?, 0x18?}, 0x64)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/parser.go:375 +0x3bf
github.com/swaggo/swag/gen.(*Gen).Build(0xc000249920, 0xc0002502a0)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/gen/gen.go:182 +0x637
main.initAction(0xc000266780?)
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/cmd/swag/main.go:158 +0x7bd
github.com/urfave/cli/v2.(*Command).Run(0xc000220ea0, 0xc000262300)
        /Users/nicolas.sassi/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/command.go:163 +0x5dc
github.com/urfave/cli/v2.(*App).RunContext(0xc000195a00, {0x165cef0?, 0xc000196008}, {0xc000192120, 0x6, 0x6})
        /Users/nicolas.sassi/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:313 +0xb7d
github.com/urfave/cli/v2.(*App).Run(...)
        /Users/nicolas.sassi/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224
main.main()
        /Users/nicolas.sassi/go/pkg/mod/github.com/swaggo/swag@v1.8.8-0.20221122034606-e5d507dd4727/cmd/swag/main.go:229 +0x5c5
make: *** [swagger] Error 2
```

====compile output====
panic: runtime error: negative shift amount

goroutine 1 [running]:
github.com/swaggo/swag.(*PackageDefinitions).evaluateConstValue(0x14000fc55e0, 0x140105e72c0?, 0x140105e81c8?, {0x1045f1298?, 0x14001646d80?}, {0x1045ef8c0, 0x14000075820}, 0x14003d8c428?)
	/Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/package.go:146 +0xdcc
github.com/swaggo/swag.(*PackagesDefinitions).EvaluateConstValue(0x1400e4bf668?, 0x14000fc55e0, 0x140101ce300, 0x0)
	/Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/packages.go:282 +0x144
github.com/swaggo/swag.(*PackagesDefinitions).evaluateAllConstVariables(...)
	/Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/packages.go:265
github.com/swaggo/swag.(*PackagesDefinitions).ParseTypes(0x14000075820)
	/Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/packages.go:110 +0x1d0
github.com/swaggo/swag.(*Parser).ParseAPIMultiSearchDir(0x140001e8000, {0x140001e2100?, 0x1?, 0x0?}, {0x10443af8a?, 0x7?}, 0x64)
	/Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/parser.go:375 +0x29c
github.com/swaggo/swag/gen.(*Gen).Build(0x140001af920, 0x140001a82a0)
	/Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/gen/gen.go:182 +0x4dc
main.initAction(0x140001d2780?)
	/Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/cmd/swag/main.go:158 +0x698
github.com/urfave/cli/v2.(*Command).Run(0x14000000c60, 0x140001d0340)
	/Users/faustfu/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/command.go:163 +0x4fc
github.com/urfave/cli/v2.(*App).RunContext(0x140000a3520, {0x1045f22b0?, 0x14000024110}, {0x140000201e0, 0x6, 0x6})
	/Users/faustfu/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:313 +0x860
github.com/urfave/cli/v2.(*App).Run(...)
	/Users/faustfu/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224
main.main()
	/Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/cmd/swag/main.go:229 +0x65c
make: *** [swag] Error 2

> ====compile output==== panic: runtime error: negative shift amount
> 
> goroutine 1 [running]: github.com/swaggo/swag.(*PackageDefinitions).evaluateConstValue(0x14000fc55e0, 0x140105e72c0?, 0x140105e81c8?, {0x1045f1298?, 0x14001646d80?}, {0x1045ef8c0, 0x14000075820}, 0x14003d8c428?) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/package.go:146 +0xdcc github.com/swaggo/swag.(*PackagesDefinitions).EvaluateConstValue(0x1400e4bf668?, 0x14000fc55e0, 0x140101ce300, 0x0) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/packages.go:282 +0x144 github.com/swaggo/swag.(*PackagesDefinitions).evaluateAllConstVariables(...) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/packages.go:265 github.com/swaggo/swag.(*PackagesDefinitions).ParseTypes(0x14000075820) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/packages.go:110 +0x1d0 github.com/swaggo/swag.(*Parser).ParseAPIMultiSearchDir(0x140001e8000, {0x140001e2100?, 0x1?, 0x0?}, {0x10443af8a?, 0x7?}, 0x64) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/parser.go:375 +0x29c github.com/swaggo/swag/gen.(*Gen).Build(0x140001af920, 0x140001a82a0) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/gen/gen.go:182 +0x4dc main.initAction(0x140001d2780?) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/cmd/swag/main.go:158 +0x698 github.com/urfave/cli/v2.(*Command).Run(0x14000000c60, 0x140001d0340) /Users/faustfu/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/command.go:163 +0x4fc github.com/urfave/cli/v2.(*App).RunContext(0x140000a3520, {0x1045f22b0?, 0x14000024110}, {0x140000201e0, 0x6, 0x6}) /Users/faustfu/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:313 +0x860 github.com/urfave/cli/v2.(*App).Run(...) /Users/faustfu/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224 main.main() /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/cmd/swag/main.go:229 +0x65c make: *** [swag] Error 2

This is not helpful to me, someone above have provided this log

> ====compile output==== panic: runtime error: negative shift amount
> 
> goroutine 1 [running]: github.com/swaggo/swag.(*PackageDefinitions).evaluateConstValue(0x14000fc55e0, 0x140105e72c0?, 0x140105e81c8?, {0x1045f1298?, 0x14001646d80?}, {0x1045ef8c0, 0x14000075820}, 0x14003d8c428?) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/package.go:146 +0xdcc github.com/swaggo/swag.(*PackagesDefinitions).EvaluateConstValue(0x1400e4bf668?, 0x14000fc55e0, 0x140101ce300, 0x0) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/packages.go:282 +0x144 github.com/swaggo/swag.(*PackagesDefinitions).evaluateAllConstVariables(...) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/packages.go:265 github.com/swaggo/swag.(*PackagesDefinitions).ParseTypes(0x14000075820) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/packages.go:110 +0x1d0 github.com/swaggo/swag.(*Parser).ParseAPIMultiSearchDir(0x140001e8000, {0x140001e2100?, 0x1?, 0x0?}, {0x10443af8a?, 0x7?}, 0x64) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/parser.go:375 +0x29c github.com/swaggo/swag/gen.(*Gen).Build(0x140001af920, 0x140001a82a0) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/gen/gen.go:182 +0x4dc main.initAction(0x140001d2780?) /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/cmd/swag/main.go:158 +0x698 github.com/urfave/cli/v2.(*Command).Run(0x14000000c60, 0x140001d0340) /Users/faustfu/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/command.go:163 +0x4fc github.com/urfave/cli/v2.(*App).RunContext(0x140000a3520, {0x1045f22b0?, 0x14000024110}, {0x140000201e0, 0x6, 0x6}) /Users/faustfu/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:313 +0x860 github.com/urfave/cli/v2.(*App).Run(...) /Users/faustfu/go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224 main.main() /Users/faustfu/go/pkg/mod/github.com/swaggo/swag@v1.8.8/cmd/swag/main.go:229 +0x65c make: *** [swag] Error 2

This is not helpful to me, someone above have provided this log

--- a/const.go
+++ b/const.go
@@ -1,13 +1,11 @@
 package swag
 
 import (
-	"fmt"
 	"go/ast"
 	"go/token"
 	"reflect"
 	"strconv"
 	"strings"
-	"unicode/utf8"
 )
 
 // ConstVariable a model to record a const variable
@@ -60,9 +58,9 @@ func EvaluateEscapedString(text string) string {
 			i++
 			if text[i] == 'u' {
 				i++
-				n, err := strconv.ParseInt(text[i:i+4], 16, 32)
+				char, err := strconv.ParseInt(text[i:i+4], 16, 32)
 				if err == nil {
-					result = utf8.AppendRune(result, rune(n))
+					result = AppendUtf8Rune(result, rune(char))
 				}
 				i += 3
 			} else if c, ok := escapedChars[text[i]]; ok {
@@ -406,7 +404,7 @@ func EvaluateUnary(x interface{}, operator token.Token, xtype ast.Expr) (interfa
 func EvaluateBinary(x, y interface{}, operator token.Token, xtype, ytype ast.Expr) (interface{}, ast.Expr) {
 	if operator == token.SHR || operator == token.SHL {
 		var rightOperand uint64
-		yValue := reflect.ValueOf(y)
+		yValue := CanIntegerValue{reflect.ValueOf(y)}
 		if yValue.CanUint() {
 			rightOperand = yValue.Uint()
 		} else if yValue.CanInt() {
@@ -469,8 +467,8 @@ func EvaluateBinary(x, y interface{}, operator token.Token, xtype, ytype ast.Exp
 		evalType = ytype
 	}
 
-	xValue := reflect.ValueOf(x)
-	yValue := reflect.ValueOf(y)
+	xValue := CanIntegerValue{reflect.ValueOf(x)}
+	yValue := CanIntegerValue{reflect.ValueOf(y)}
 	if xValue.Kind() == reflect.String && yValue.Kind() == reflect.String {
 		return xValue.String() + yValue.String(), evalType
 	}
@@ -565,276 +563,4 @@ func EvaluateBinary(x, y interface{}, operator token.Token, xtype, ytype ast.Exp
 		}
 	}
 	return targetValue.Interface(), evalType
-
-	switch operator {
-	case token.ADD:
-		switch xValue := x.(type) {
-		case int:
-			switch yValue := y.(type) {
-			case int:
-				return xValue + yValue, evalType
-			case int8:
-				return int8(xValue) + yValue, evalType
-			case int16:
-				return int16(xValue) + yValue, evalType
-			case int32:
-				return int32(xValue) + yValue, evalType
-			case int64:
-				return int64(xValue) + yValue, evalType
-			case uint:
-				return uint(xValue) + yValue, evalType
-			case uint8:
-				return uint8(xValue) + yValue, evalType
-			case uint16:
-				return uint16(xValue) + yValue, evalType
-			case uint32:
-				return uint32(xValue) + yValue, evalType
-			case uint64:
-				return uint64(xValue) + yValue, evalType
-			}
-			return xValue + y.(int), evalType
-		case int8:
-			switch yValue := y.(type) {
-			case int:
-				return xValue + int8(yValue), evalType
-			case int8:
-				return xValue + yValue, evalType
-			}
-			return xValue + y.(int8), evalType
-		case int16:
-			return xValue + y.(int16), evalType
-		case int32:
-			return xValue + y.(int32), evalType
-		case int64:
-			return xValue + y.(int64), evalType
-		case uint:
-			return xValue + y.(uint), evalType
-		case uint8:
-			return xValue + y.(uint8), evalType
-		case uint16:
-			return xValue + y.(uint16), evalType
-		case uint32:
-			return xValue + y.(uint32), evalType
-		case uint64:
-			return xValue + y.(uint64), evalType
-		case string:
-			return xValue + y.(string), evalType
-		}
-	case token.SUB:
-		switch xValue := x.(type) {
-		case int:
-			return xValue - y.(int), evalType
-		case int8:
-			return xValue - y.(int8), evalType
-		case int16:
-			return xValue - y.(int16), evalType
-		case int32:
-			return xValue - y.(int32), evalType
-		case int64:
-			return xValue - y.(int64), evalType
-		case uint:
-			return xValue - y.(uint), evalType
-		case uint8:
-			return xValue - y.(uint8), evalType
-		case uint16:
-			return xValue - y.(uint16), evalType
-		case uint32:
-			return xValue - y.(uint32), evalType
-		case uint64:
-			return xValue - y.(uint64), evalType
-		}
-	case token.MUL:
-		switch xValue := x.(type) {
-		case int:
-			return xValue * y.(int), evalType
-		case int8:
-			return xValue * y.(int8), evalType
-		case int16:
-			return xValue * y.(int16), evalType
-		case int32:
-			return xValue * y.(int32), evalType
-		case int64:
-			return xValue * y.(int64), evalType
-		case uint:
-			return xValue * y.(uint), evalType
-		case uint8:
-			return xValue * y.(uint8), evalType
-		case uint16:
-			return xValue * y.(uint16), evalType
-		case uint32:
-			return xValue * y.(uint32), evalType
-		case uint64:
-			return xValue * y.(uint64), evalType
-		}
-	case token.QUO:
-		switch xValue := x.(type) {
-		case int:
-			return xValue / y.(int), evalType
-		case int8:
-			return xValue / y.(int8), evalType
-		case int16:
-			return xValue / y.(int16), evalType
-		case int32:
-			return xValue / y.(int32), evalType
-		case int64:
-			return xValue / y.(int64), evalType
-		case uint:
-			return xValue / y.(uint), evalType
-		case uint8:
-			return xValue / y.(uint8), evalType
-		case uint16:
-			return xValue / y.(uint16), evalType
-		case uint32:
-			return xValue / y.(uint32), evalType
-		case uint64:
-			return xValue / y.(uint64), evalType
-		}
-	case token.REM:
-		switch xValue := x.(type) {
-		case int:
-			return xValue % y.(int), evalType
-		case int8:
-			return xValue % y.(int8), evalType
-		case int16:
-			return xValue % y.(int16), evalType
-		case int32:
-			return xValue % y.(int32), evalType
-		case int64:
-			return xValue % y.(int64), evalType
-		case uint:
-			return xValue % y.(uint), evalType
-		case uint8:
-			return xValue % y.(uint8), evalType
-		case uint16:
-			return xValue % y.(uint16), evalType
-		case uint32:
-			return xValue % y.(uint32), evalType
-		case uint64:
-			return xValue % y.(uint64), evalType
-		}
-	case token.AND:
-		switch xValue := x.(type) {
-		case int:
-			return xValue & y.(int), evalType
-		case int8:
-			return xValue & y.(int8), evalType
-		case int16:
-			return xValue & y.(int16), evalType
-		case int32:
-			return xValue & y.(int32), evalType
-		case int64:
-			return xValue & y.(int64), evalType
-		case uint:
-			return xValue & y.(uint), evalType
-		case uint8:
-			return xValue & y.(uint8), evalType
-		case uint16:
-			return xValue & y.(uint16), evalType
-		case uint32:
-			return xValue & y.(uint32), evalType
-		case uint64:
-			return xValue & y.(uint64), evalType
-		}
-	case token.OR:
-		switch xValue := x.(type) {
-		case int:
-			return xValue | y.(int), evalType
-		case int8:
-			return xValue | y.(int8), evalType
-		case int16:
-			return xValue | y.(int16), evalType
-		case int32:
-			return xValue | y.(int32), evalType
-		case int64:
-			return xValue | y.(int64), evalType
-		case uint:
-			return xValue | y.(uint), evalType
-		case uint8:
-			return xValue | y.(uint8), evalType
-		case uint16:
-			return xValue | y.(uint16), evalType
-		case uint32:
-			return xValue | y.(uint32), evalType
-		case uint64:
-			return xValue | y.(uint64), evalType
-		}
-	case token.XOR:
-		switch xValue := x.(type) {
-		case int:
-			return xValue ^ y.(int), evalType
-		case int8:
-			return xValue ^ y.(int8), evalType
-		case int16:
-			return xValue ^ y.(int16), evalType
-		case int32:
-			return xValue ^ y.(int32), evalType
-		case int64:
-			return xValue ^ y.(int64), evalType
-		case uint:
-			return xValue ^ y.(uint), evalType
-		case uint8:
-			return xValue ^ y.(uint8), evalType
-		case uint16:
-			return xValue ^ y.(uint16), evalType
-		case uint32:
-			return xValue ^ y.(uint32), evalType
-		case uint64:
-			return xValue ^ y.(uint64), evalType
-		}
-	case token.SHL:
-		rightOperand, err := strconv.ParseUint(fmt.Sprintf("%v", y), 10, 64)
-		if err != nil {
-			panic(err)
-		}
-		switch xValue := x.(type) {
-		case int:
-			return xValue << rightOperand, xtype
-		case int8:
-			return xValue << rightOperand, xtype
-		case int16:
-			return xValue << rightOperand, xtype
-		case int32:
-			return xValue << rightOperand, xtype
-		case int64:
-			return xValue << rightOperand, xtype
-		case uint:
-			return xValue << rightOperand, xtype
-		case uint8:
-			return xValue << rightOperand, xtype
-		case uint16:
-			return xValue << rightOperand, xtype
-		case uint32:
-			return xValue << rightOperand, xtype
-		case uint64:
-			return xValue << rightOperand, xtype
-		}
-	case token.SHR:
-		rightOperand, err := strconv.ParseUint(fmt.Sprintf("%v", y), 10, 64)
-		if err != nil {
-			panic(err)
-		}
-		switch xValue := x.(type) {
-		case int:
-			return xValue >> rightOperand, xtype
-		case int8:
-			return xValue >> rightOperand, xtype
-		case int16:
-			return xValue >> rightOperand, xtype
-		case int32:
-			return xValue >> rightOperand, xtype
-		case int64:
-			return xValue >> rightOperand, xtype
-		case uint:
-			return xValue >> rightOperand, xtype
-		case uint8:
-			return xValue >> rightOperand, xtype
-		case uint16:
-			return xValue >> rightOperand, xtype
-		case uint32:
-			return xValue >> rightOperand, xtype
-		case uint64:
-			return xValue >> rightOperand, xtype
-		}
-	}
-	return nil, nil
 }

--- a/const.go
+++ b/const.go
@@ -29,6 +29,7 @@ var escapedChars = map[uint8]uint8{
 	'"':  '"',
 }
 
+// EvaluateEscapedChar parse escaped character
 func EvaluateEscapedChar(text string) rune {
 	if len(text) == 1 {
 		return rune(text[0])
@@ -48,6 +49,7 @@ func EvaluateEscapedChar(text string) rune {
 	return 0
 }
 
+// EvaluateEscapedString parse escaped characters in string
 func EvaluateEscapedString(text string) string {
 	if !strings.ContainsRune(text, '\\') {
 		return text
@@ -73,6 +75,7 @@ func EvaluateEscapedString(text string) string {
 	return string(result)
 }
 
+// EvaluateDataConversion evaluate the type a explicit type conversion
 func EvaluateDataConversion(x interface{}, typeName string) interface{} {
 	switch value := x.(type) {
 	case int:
@@ -356,6 +359,7 @@ func EvaluateDataConversion(x interface{}, typeName string) interface{} {
 	return nil
 }
 
+// EvaluateUnary evaluate the type and value of a unary expression
 func EvaluateUnary(x interface{}, operator token.Token, xtype ast.Expr) (interface{}, ast.Expr) {
 	switch operator {
 	case token.SUB:
@@ -398,6 +402,7 @@ func EvaluateUnary(x interface{}, operator token.Token, xtype ast.Expr) (interfa
 	return nil, nil
 }
 
+// EvaluateBinary evaluate the type and value of a binary expression
 func EvaluateBinary(x, y interface{}, operator token.Token, xtype, ytype ast.Expr) (interface{}, ast.Expr) {
 	if operator == token.SHR || operator == token.SHL {
 		var rightOperand uint64

--- a/const.go
+++ b/const.go
@@ -1,7 +1,13 @@
 package swag
 
 import (
+	"fmt"
 	"go/ast"
+	"go/token"
+	"reflect"
+	"strconv"
+	"strings"
+	"unicode/utf8"
 )
 
 // ConstVariable a model to record a const variable
@@ -12,4 +18,818 @@ type ConstVariable struct {
 	Comment *ast.CommentGroup
 	File    *ast.File
 	Pkg     *PackageDefinitions
+}
+
+var escapedChars = map[uint8]uint8{
+	'n':  '\n',
+	'r':  '\r',
+	't':  '\t',
+	'v':  '\v',
+	'\\': '\\',
+	'"':  '"',
+}
+
+func EvaluateEscapedChar(text string) rune {
+	if len(text) == 1 {
+		return rune(text[0])
+	}
+
+	if len(text) == 2 && text[0] == '\\' {
+		return rune(escapedChars[text[1]])
+	}
+
+	if len(text) == 6 && text[0:2] == "\\u" {
+		n, err := strconv.ParseInt(text[2:], 16, 32)
+		if err == nil {
+			return rune(n)
+		}
+	}
+
+	return 0
+}
+
+func EvaluateEscapedString(text string) string {
+	if !strings.ContainsRune(text, '\\') {
+		return text
+	}
+	result := make([]byte, 0, len(text))
+	for i := 0; i < len(text); i++ {
+		if text[i] == '\\' {
+			i++
+			if text[i] == 'u' {
+				i++
+				n, err := strconv.ParseInt(text[i:i+4], 16, 32)
+				if err == nil {
+					result = utf8.AppendRune(result, rune(n))
+				}
+				i += 3
+			} else if c, ok := escapedChars[text[i]]; ok {
+				result = append(result, c)
+			}
+		} else {
+			result = append(result, text[i])
+		}
+	}
+	return string(result)
+}
+
+func EvaluateDataConversion(x interface{}, typeName string) interface{} {
+	switch value := x.(type) {
+	case int:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		}
+	case uint:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		}
+	case int8:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		}
+	case uint8:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		}
+	case int16:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		}
+	case uint16:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		}
+	case int32:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		case "string":
+			return string(value)
+		}
+	case uint32:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		}
+	case int64:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		}
+	case uint64:
+		switch typeName {
+		case "int":
+			return int(value)
+		case "byte":
+			return byte(value)
+		case "int8":
+			return int8(value)
+		case "int16":
+			return int16(value)
+		case "int32":
+			return int32(value)
+		case "int64":
+			return int64(value)
+		case "uint":
+			return uint(value)
+		case "uint8":
+			return uint8(value)
+		case "uint16":
+			return uint16(value)
+		case "uint32":
+			return uint32(value)
+		case "uint64":
+			return uint64(value)
+		case "rune":
+			return rune(value)
+		}
+	case string:
+		switch typeName {
+		case "string":
+			return value
+		}
+	}
+	return nil
+}
+
+func EvaluateUnary(x interface{}, operator token.Token, xtype ast.Expr) (interface{}, ast.Expr) {
+	switch operator {
+	case token.SUB:
+		switch value := x.(type) {
+		case int:
+			return -value, xtype
+		case int8:
+			return -value, xtype
+		case int16:
+			return -value, xtype
+		case int32:
+			return -value, xtype
+		case int64:
+			return -value, xtype
+		}
+	case token.XOR:
+		switch value := x.(type) {
+		case int:
+			return ^value, xtype
+		case int8:
+			return ^value, xtype
+		case int16:
+			return ^value, xtype
+		case int32:
+			return ^value, xtype
+		case int64:
+			return ^value, xtype
+		case uint:
+			return ^value, xtype
+		case uint8:
+			return ^value, xtype
+		case uint16:
+			return ^value, xtype
+		case uint32:
+			return ^value, xtype
+		case uint64:
+			return ^value, xtype
+		}
+	}
+	return nil, nil
+}
+
+func EvaluateBinary(x, y interface{}, operator token.Token, xtype, ytype ast.Expr) (interface{}, ast.Expr) {
+	if operator == token.SHR || operator == token.SHL {
+		var rightOperand uint64
+		yValue := reflect.ValueOf(y)
+		if yValue.CanUint() {
+			rightOperand = yValue.Uint()
+		} else if yValue.CanInt() {
+			rightOperand = uint64(yValue.Int())
+		}
+
+		switch operator {
+		case token.SHL:
+			switch xValue := x.(type) {
+			case int:
+				return xValue << rightOperand, xtype
+			case int8:
+				return xValue << rightOperand, xtype
+			case int16:
+				return xValue << rightOperand, xtype
+			case int32:
+				return xValue << rightOperand, xtype
+			case int64:
+				return xValue << rightOperand, xtype
+			case uint:
+				return xValue << rightOperand, xtype
+			case uint8:
+				return xValue << rightOperand, xtype
+			case uint16:
+				return xValue << rightOperand, xtype
+			case uint32:
+				return xValue << rightOperand, xtype
+			case uint64:
+				return xValue << rightOperand, xtype
+			}
+		case token.SHR:
+			switch xValue := x.(type) {
+			case int:
+				return xValue >> rightOperand, xtype
+			case int8:
+				return xValue >> rightOperand, xtype
+			case int16:
+				return xValue >> rightOperand, xtype
+			case int32:
+				return xValue >> rightOperand, xtype
+			case int64:
+				return xValue >> rightOperand, xtype
+			case uint:
+				return xValue >> rightOperand, xtype
+			case uint8:
+				return xValue >> rightOperand, xtype
+			case uint16:
+				return xValue >> rightOperand, xtype
+			case uint32:
+				return xValue >> rightOperand, xtype
+			case uint64:
+				return xValue >> rightOperand, xtype
+			}
+		}
+		return nil, nil
+	}
+
+	evalType := xtype
+	if evalType == nil {
+		evalType = ytype
+	}
+
+	xValue := reflect.ValueOf(x)
+	yValue := reflect.ValueOf(y)
+	if xValue.Kind() == reflect.String && yValue.Kind() == reflect.String {
+		return xValue.String() + yValue.String(), evalType
+	}
+
+	var targetValue reflect.Value
+	if xValue.Kind() != reflect.Int {
+		targetValue = reflect.New(xValue.Type()).Elem()
+	} else {
+		targetValue = reflect.New(yValue.Type()).Elem()
+	}
+
+	switch operator {
+	case token.ADD:
+		if xValue.CanInt() && yValue.CanInt() {
+			targetValue.SetInt(xValue.Int() + yValue.Int())
+		} else if xValue.CanUint() && yValue.CanUint() {
+			targetValue.SetUint(xValue.Uint() + yValue.Uint())
+		} else if xValue.CanInt() && yValue.CanUint() {
+			targetValue.SetUint(uint64(xValue.Int()) + yValue.Uint())
+		} else if xValue.CanUint() && yValue.CanInt() {
+			targetValue.SetUint(xValue.Uint() + uint64(yValue.Int()))
+		}
+	case token.SUB:
+		if xValue.CanInt() && yValue.CanInt() {
+			targetValue.SetInt(xValue.Int() - yValue.Int())
+		} else if xValue.CanUint() && yValue.CanUint() {
+			targetValue.SetUint(xValue.Uint() - yValue.Uint())
+		} else if xValue.CanInt() && yValue.CanUint() {
+			targetValue.SetUint(uint64(xValue.Int()) - yValue.Uint())
+		} else if xValue.CanUint() && yValue.CanInt() {
+			targetValue.SetUint(xValue.Uint() - uint64(yValue.Int()))
+		}
+	case token.MUL:
+		if xValue.CanInt() && yValue.CanInt() {
+			targetValue.SetInt(xValue.Int() * yValue.Int())
+		} else if xValue.CanUint() && yValue.CanUint() {
+			targetValue.SetUint(xValue.Uint() * yValue.Uint())
+		} else if xValue.CanInt() && yValue.CanUint() {
+			targetValue.SetUint(uint64(xValue.Int()) * yValue.Uint())
+		} else if xValue.CanUint() && yValue.CanInt() {
+			targetValue.SetUint(xValue.Uint() * uint64(yValue.Int()))
+		}
+	case token.QUO:
+		if xValue.CanInt() && yValue.CanInt() {
+			targetValue.SetInt(xValue.Int() / yValue.Int())
+		} else if xValue.CanUint() && yValue.CanUint() {
+			targetValue.SetUint(xValue.Uint() / yValue.Uint())
+		} else if xValue.CanInt() && yValue.CanUint() {
+			targetValue.SetUint(uint64(xValue.Int()) / yValue.Uint())
+		} else if xValue.CanUint() && yValue.CanInt() {
+			targetValue.SetUint(xValue.Uint() / uint64(yValue.Int()))
+		}
+	case token.REM:
+		if xValue.CanInt() && yValue.CanInt() {
+			targetValue.SetInt(xValue.Int() % yValue.Int())
+		} else if xValue.CanUint() && yValue.CanUint() {
+			targetValue.SetUint(xValue.Uint() % yValue.Uint())
+		} else if xValue.CanInt() && yValue.CanUint() {
+			targetValue.SetUint(uint64(xValue.Int()) % yValue.Uint())
+		} else if xValue.CanUint() && yValue.CanInt() {
+			targetValue.SetUint(xValue.Uint() % uint64(yValue.Int()))
+		}
+	case token.AND:
+		if xValue.CanInt() && yValue.CanInt() {
+			targetValue.SetInt(xValue.Int() & yValue.Int())
+		} else if xValue.CanUint() && yValue.CanUint() {
+			targetValue.SetUint(xValue.Uint() & yValue.Uint())
+		} else if xValue.CanInt() && yValue.CanUint() {
+			targetValue.SetUint(uint64(xValue.Int()) & yValue.Uint())
+		} else if xValue.CanUint() && yValue.CanInt() {
+			targetValue.SetUint(xValue.Uint() & uint64(yValue.Int()))
+		}
+	case token.OR:
+		if xValue.CanInt() && yValue.CanInt() {
+			targetValue.SetInt(xValue.Int() | yValue.Int())
+		} else if xValue.CanUint() && yValue.CanUint() {
+			targetValue.SetUint(xValue.Uint() | yValue.Uint())
+		} else if xValue.CanInt() && yValue.CanUint() {
+			targetValue.SetUint(uint64(xValue.Int()) | yValue.Uint())
+		} else if xValue.CanUint() && yValue.CanInt() {
+			targetValue.SetUint(xValue.Uint() | uint64(yValue.Int()))
+		}
+	case token.XOR:
+		if xValue.CanInt() && yValue.CanInt() {
+			targetValue.SetInt(xValue.Int() ^ yValue.Int())
+		} else if xValue.CanUint() && yValue.CanUint() {
+			targetValue.SetUint(xValue.Uint() ^ yValue.Uint())
+		} else if xValue.CanInt() && yValue.CanUint() {
+			targetValue.SetUint(uint64(xValue.Int()) ^ yValue.Uint())
+		} else if xValue.CanUint() && yValue.CanInt() {
+			targetValue.SetUint(xValue.Uint() ^ uint64(yValue.Int()))
+		}
+	}
+	return targetValue.Interface(), evalType
+
+	switch operator {
+	case token.ADD:
+		switch xValue := x.(type) {
+		case int:
+			switch yValue := y.(type) {
+			case int:
+				return xValue + yValue, evalType
+			case int8:
+				return int8(xValue) + yValue, evalType
+			case int16:
+				return int16(xValue) + yValue, evalType
+			case int32:
+				return int32(xValue) + yValue, evalType
+			case int64:
+				return int64(xValue) + yValue, evalType
+			case uint:
+				return uint(xValue) + yValue, evalType
+			case uint8:
+				return uint8(xValue) + yValue, evalType
+			case uint16:
+				return uint16(xValue) + yValue, evalType
+			case uint32:
+				return uint32(xValue) + yValue, evalType
+			case uint64:
+				return uint64(xValue) + yValue, evalType
+			}
+			return xValue + y.(int), evalType
+		case int8:
+			switch yValue := y.(type) {
+			case int:
+				return xValue + int8(yValue), evalType
+			case int8:
+				return xValue + yValue, evalType
+			}
+			return xValue + y.(int8), evalType
+		case int16:
+			return xValue + y.(int16), evalType
+		case int32:
+			return xValue + y.(int32), evalType
+		case int64:
+			return xValue + y.(int64), evalType
+		case uint:
+			return xValue + y.(uint), evalType
+		case uint8:
+			return xValue + y.(uint8), evalType
+		case uint16:
+			return xValue + y.(uint16), evalType
+		case uint32:
+			return xValue + y.(uint32), evalType
+		case uint64:
+			return xValue + y.(uint64), evalType
+		case string:
+			return xValue + y.(string), evalType
+		}
+	case token.SUB:
+		switch xValue := x.(type) {
+		case int:
+			return xValue - y.(int), evalType
+		case int8:
+			return xValue - y.(int8), evalType
+		case int16:
+			return xValue - y.(int16), evalType
+		case int32:
+			return xValue - y.(int32), evalType
+		case int64:
+			return xValue - y.(int64), evalType
+		case uint:
+			return xValue - y.(uint), evalType
+		case uint8:
+			return xValue - y.(uint8), evalType
+		case uint16:
+			return xValue - y.(uint16), evalType
+		case uint32:
+			return xValue - y.(uint32), evalType
+		case uint64:
+			return xValue - y.(uint64), evalType
+		}
+	case token.MUL:
+		switch xValue := x.(type) {
+		case int:
+			return xValue * y.(int), evalType
+		case int8:
+			return xValue * y.(int8), evalType
+		case int16:
+			return xValue * y.(int16), evalType
+		case int32:
+			return xValue * y.(int32), evalType
+		case int64:
+			return xValue * y.(int64), evalType
+		case uint:
+			return xValue * y.(uint), evalType
+		case uint8:
+			return xValue * y.(uint8), evalType
+		case uint16:
+			return xValue * y.(uint16), evalType
+		case uint32:
+			return xValue * y.(uint32), evalType
+		case uint64:
+			return xValue * y.(uint64), evalType
+		}
+	case token.QUO:
+		switch xValue := x.(type) {
+		case int:
+			return xValue / y.(int), evalType
+		case int8:
+			return xValue / y.(int8), evalType
+		case int16:
+			return xValue / y.(int16), evalType
+		case int32:
+			return xValue / y.(int32), evalType
+		case int64:
+			return xValue / y.(int64), evalType
+		case uint:
+			return xValue / y.(uint), evalType
+		case uint8:
+			return xValue / y.(uint8), evalType
+		case uint16:
+			return xValue / y.(uint16), evalType
+		case uint32:
+			return xValue / y.(uint32), evalType
+		case uint64:
+			return xValue / y.(uint64), evalType
+		}
+	case token.REM:
+		switch xValue := x.(type) {
+		case int:
+			return xValue % y.(int), evalType
+		case int8:
+			return xValue % y.(int8), evalType
+		case int16:
+			return xValue % y.(int16), evalType
+		case int32:
+			return xValue % y.(int32), evalType
+		case int64:
+			return xValue % y.(int64), evalType
+		case uint:
+			return xValue % y.(uint), evalType
+		case uint8:
+			return xValue % y.(uint8), evalType
+		case uint16:
+			return xValue % y.(uint16), evalType
+		case uint32:
+			return xValue % y.(uint32), evalType
+		case uint64:
+			return xValue % y.(uint64), evalType
+		}
+	case token.AND:
+		switch xValue := x.(type) {
+		case int:
+			return xValue & y.(int), evalType
+		case int8:
+			return xValue & y.(int8), evalType
+		case int16:
+			return xValue & y.(int16), evalType
+		case int32:
+			return xValue & y.(int32), evalType
+		case int64:
+			return xValue & y.(int64), evalType
+		case uint:
+			return xValue & y.(uint), evalType
+		case uint8:
+			return xValue & y.(uint8), evalType
+		case uint16:
+			return xValue & y.(uint16), evalType
+		case uint32:
+			return xValue & y.(uint32), evalType
+		case uint64:
+			return xValue & y.(uint64), evalType
+		}
+	case token.OR:
+		switch xValue := x.(type) {
+		case int:
+			return xValue | y.(int), evalType
+		case int8:
+			return xValue | y.(int8), evalType
+		case int16:
+			return xValue | y.(int16), evalType
+		case int32:
+			return xValue | y.(int32), evalType
+		case int64:
+			return xValue | y.(int64), evalType
+		case uint:
+			return xValue | y.(uint), evalType
+		case uint8:
+			return xValue | y.(uint8), evalType
+		case uint16:
+			return xValue | y.(uint16), evalType
+		case uint32:
+			return xValue | y.(uint32), evalType
+		case uint64:
+			return xValue | y.(uint64), evalType
+		}
+	case token.XOR:
+		switch xValue := x.(type) {
+		case int:
+			return xValue ^ y.(int), evalType
+		case int8:
+			return xValue ^ y.(int8), evalType
+		case int16:
+			return xValue ^ y.(int16), evalType
+		case int32:
+			return xValue ^ y.(int32), evalType
+		case int64:
+			return xValue ^ y.(int64), evalType
+		case uint:
+			return xValue ^ y.(uint), evalType
+		case uint8:
+			return xValue ^ y.(uint8), evalType
+		case uint16:
+			return xValue ^ y.(uint16), evalType
+		case uint32:
+			return xValue ^ y.(uint32), evalType
+		case uint64:
+			return xValue ^ y.(uint64), evalType
+		}
+	case token.SHL:
+		rightOperand, err := strconv.ParseUint(fmt.Sprintf("%v", y), 10, 64)
+		if err != nil {
+			panic(err)
+		}
+		switch xValue := x.(type) {
+		case int:
+			return xValue << rightOperand, xtype
+		case int8:
+			return xValue << rightOperand, xtype
+		case int16:
+			return xValue << rightOperand, xtype
+		case int32:
+			return xValue << rightOperand, xtype
+		case int64:
+			return xValue << rightOperand, xtype
+		case uint:
+			return xValue << rightOperand, xtype
+		case uint8:
+			return xValue << rightOperand, xtype
+		case uint16:
+			return xValue << rightOperand, xtype
+		case uint32:
+			return xValue << rightOperand, xtype
+		case uint64:
+			return xValue << rightOperand, xtype
+		}
+	case token.SHR:
+		rightOperand, err := strconv.ParseUint(fmt.Sprintf("%v", y), 10, 64)
+		if err != nil {
+			panic(err)
+		}
+		switch xValue := x.(type) {
+		case int:
+			return xValue >> rightOperand, xtype
+		case int8:
+			return xValue >> rightOperand, xtype
+		case int16:
+			return xValue >> rightOperand, xtype
+		case int32:
+			return xValue >> rightOperand, xtype
+		case int64:
+			return xValue >> rightOperand, xtype
+		case uint:
+			return xValue >> rightOperand, xtype
+		case uint8:
+			return xValue >> rightOperand, xtype
+		case uint16:
+			return xValue >> rightOperand, xtype
+		case uint32:
+			return xValue >> rightOperand, xtype
+		case uint64:
+			return xValue >> rightOperand, xtype
+		}
+	}
+	return nil, nil
 }

--- a/enums_test.go
+++ b/enums_test.go
@@ -20,4 +20,13 @@ func TestParseGlobalEnums(t *testing.T) {
 	b, err := json.MarshalIndent(p.swagger, "", "    ")
 	assert.NoError(t, err)
 	assert.Equal(t, string(expected), string(b))
+	constsPath := "github.com/swaggo/swag/testdata/enums/consts"
+	assert.Equal(t, 64, p.packages.packages[constsPath].ConstTable["uintSize"].Value)
+	assert.Equal(t, int32(62), p.packages.packages[constsPath].ConstTable["maxBase"].Value)
+	assert.Equal(t, 8, p.packages.packages[constsPath].ConstTable["shlByLen"].Value)
+	assert.Equal(t, 255, p.packages.packages[constsPath].ConstTable["hexnum"].Value)
+	assert.Equal(t, 15, p.packages.packages[constsPath].ConstTable["octnum"].Value)
+	assert.Equal(t, `aa\nbb\u8888cc`, p.packages.packages[constsPath].ConstTable["nonescapestr"].Value)
+	assert.Equal(t, "aa\nbb\u8888cc", p.packages.packages[constsPath].ConstTable["escapestr"].Value)
+	assert.Equal(t, '\u8888', p.packages.packages[constsPath].ConstTable["escapechar"].Value)
 }

--- a/package.go
+++ b/package.go
@@ -3,6 +3,7 @@ package swag
 import (
 	"go/ast"
 	"go/token"
+	"reflect"
 	"strconv"
 )
 
@@ -31,6 +32,7 @@ type PackageDefinitions struct {
 type ConstVariableGlobalEvaluator interface {
 	EvaluateConstValue(pkg *PackageDefinitions, cv *ConstVariable, recursiveStack map[string]struct{}) (interface{}, ast.Expr)
 	EvaluateConstValueByName(file *ast.File, pkgPath, constVariableName string, recursiveStack map[string]struct{}) (interface{}, ast.Expr)
+	FindTypeSpec(typeName string, file *ast.File) *TypeSpecDef
 }
 
 // NewPackageDefinitions new a PackageDefinitions object
@@ -92,68 +94,89 @@ func (pkg *PackageDefinitions) evaluateConstValue(file *ast.File, iota int, expr
 	case *ast.BasicLit:
 		switch valueExpr.Kind {
 		case token.INT:
-			x, err := strconv.ParseInt(valueExpr.Value, 10, 64)
-			if err != nil {
-				return nil, nil
+			// hexadecimal
+			if len(valueExpr.Value) > 2 && valueExpr.Value[0] == '0' && valueExpr.Value[1] == 'x' {
+				if x, err := strconv.ParseInt(valueExpr.Value[2:], 16, 64); err == nil {
+					return int(x), nil
+				} else if x, err := strconv.ParseUint(valueExpr.Value[2:], 16, 64); err == nil {
+					return x, nil
+				} else {
+					panic(err)
+				}
 			}
-			return int(x), nil
-		case token.STRING, token.CHAR:
-			return valueExpr.Value[1 : len(valueExpr.Value)-1], nil
+
+			//octet
+			if len(valueExpr.Value) > 1 && valueExpr.Value[0] == '0' {
+				if x, err := strconv.ParseInt(valueExpr.Value[1:], 8, 64); err == nil {
+					return int(x), nil
+				} else if x, err := strconv.ParseUint(valueExpr.Value[1:], 8, 64); err == nil {
+					return x, nil
+				} else {
+					panic(err)
+				}
+			}
+
+			//a basic literal integer is int type in default, or must have an explicit converting type in front
+			if x, err := strconv.ParseInt(valueExpr.Value, 10, 64); err == nil {
+				return int(x), nil
+			} else if x, err := strconv.ParseUint(valueExpr.Value, 10, 64); err == nil {
+				return x, nil
+			} else {
+				panic(err)
+			}
+		case token.STRING:
+			if valueExpr.Value[0] == '`' {
+				return valueExpr.Value[1 : len(valueExpr.Value)-1], nil
+			}
+			return EvaluateEscapedString(valueExpr.Value[1 : len(valueExpr.Value)-1]), nil
+		case token.CHAR:
+			return EvaluateEscapedChar(valueExpr.Value[1 : len(valueExpr.Value)-1]), nil
 		}
 	case *ast.UnaryExpr:
 		x, evalType := pkg.evaluateConstValue(file, iota, valueExpr.X, globalEvaluator, recursiveStack)
 		if x == nil {
-			return nil, nil
+			return x, evalType
 		}
-		switch valueExpr.Op {
-		case token.SUB:
-			return -x.(int), evalType
-		case token.XOR:
-			return ^(x.(int)), evalType
-		}
+		return EvaluateUnary(x, valueExpr.Op, evalType)
 	case *ast.BinaryExpr:
 		x, evalTypex := pkg.evaluateConstValue(file, iota, valueExpr.X, globalEvaluator, recursiveStack)
 		y, evalTypey := pkg.evaluateConstValue(file, iota, valueExpr.Y, globalEvaluator, recursiveStack)
 		if x == nil || y == nil {
 			return nil, nil
 		}
-		evalType := evalTypex
-		if evalType == nil {
-			evalType = evalTypey
-		}
-		switch valueExpr.Op {
-		case token.ADD:
-			if ix, ok := x.(int); ok {
-				return ix + y.(int), evalType
-			} else if sx, ok := x.(string); ok {
-				return sx + y.(string), evalType
-			}
-		case token.SUB:
-			return x.(int) - y.(int), evalType
-		case token.MUL:
-			return x.(int) * y.(int), evalType
-		case token.QUO:
-			return x.(int) / y.(int), evalType
-		case token.REM:
-			return x.(int) % y.(int), evalType
-		case token.AND:
-			return x.(int) & y.(int), evalType
-		case token.OR:
-			return x.(int) | y.(int), evalType
-		case token.XOR:
-			return x.(int) ^ y.(int), evalType
-		case token.SHL:
-			return x.(int) << y.(int), evalType
-		case token.SHR:
-			return x.(int) >> y.(int), evalType
-		}
+		return EvaluateBinary(x, y, valueExpr.Op, evalTypex, evalTypey)
 	case *ast.ParenExpr:
 		return pkg.evaluateConstValue(file, iota, valueExpr.X, globalEvaluator, recursiveStack)
 	case *ast.CallExpr:
 		//data conversion
-		if ident, ok := valueExpr.Fun.(*ast.Ident); ok && len(valueExpr.Args) == 1 && IsGolangPrimitiveType(ident.Name) {
-			arg, _ := pkg.evaluateConstValue(file, iota, valueExpr.Args[0], globalEvaluator, recursiveStack)
-			return arg, nil
+		if len(valueExpr.Args) != 1 {
+			return nil, nil
+		}
+		arg := valueExpr.Args[0]
+		if ident, ok := valueExpr.Fun.(*ast.Ident); ok {
+			name := ident.Name
+			if name == "uintptr" {
+				name = "uint"
+			}
+			if IsGolangPrimitiveType(name) {
+				value, _ := pkg.evaluateConstValue(file, iota, arg, globalEvaluator, recursiveStack)
+				value = EvaluateDataConversion(value, name)
+				return value, nil
+			} else if name == "len" {
+				value, _ := pkg.evaluateConstValue(file, iota, arg, globalEvaluator, recursiveStack)
+				return reflect.ValueOf(value).Len(), nil
+			}
+			typeDef := globalEvaluator.FindTypeSpec(name, file)
+			if typeDef == nil {
+				return nil, nil
+			}
+			return arg, valueExpr.Fun
+		} else if selector, ok := valueExpr.Fun.(*ast.SelectorExpr); ok {
+			typeDef := globalEvaluator.FindTypeSpec(fullTypeName(selector.X.(*ast.Ident).Name, selector.Sel.Name), file)
+			if typeDef == nil {
+				return nil, nil
+			}
+			return arg, typeDef.TypeSpec.Type
 		}
 	}
 	return nil, nil

--- a/testdata/enums/consts/const.go
+++ b/testdata/enums/consts/const.go
@@ -1,3 +1,12 @@
 package consts
 
 const Base = 1
+
+const uintSize = 32 << (^uint(uintptr(0)) >> 63)
+const maxBase = 10 + ('z' - 'a' + 1) + ('Z' - 'A' + 1)
+const shlByLen = 1 << len("aaa")
+const hexnum = 0xFF
+const octnum = 017
+const nonescapestr = `aa\nbb\u8888cc`
+const escapestr = "aa\nbb\u8888cc"
+const escapechar = '\u8888'

--- a/testdata/enums/main.go
+++ b/testdata/enums/main.go
@@ -14,5 +14,4 @@ package main
 
 // @BasePath  /v2
 func main() {
-
 }

--- a/testdata/enums/types/model.go
+++ b/testdata/enums/types/model.go
@@ -11,8 +11,8 @@ const (
 	A    Class = consts.Base + (iota+1-1)*2/2%100 - (1&1 | 1) + (2 ^ 2) // AAA
 	B                                                                   /* BBB */
 	C
-	D
-	F = D + 1
+	D = C + 1
+	F = Class(5)
 	//G is not enum
 	G = H + 10
 	//H is not enum
@@ -21,13 +21,15 @@ const (
 	I = int(F + 2)
 )
 
+const J = 1 << uint16(I)
+
 type Mask int
 
 const (
-	Mask1 Mask = 2 << iota >> 1 // Mask1
-	Mask2                       /* Mask2 */
-	Mask3                       // Mask3
-	Mask4                       // Mask4
+	Mask1 Mask = 0x02 << iota >> 1 // Mask1
+	Mask2                          /* Mask2 */
+	Mask3                          // Mask3
+	Mask4                          // Mask4
 )
 
 type Type string
@@ -38,6 +40,13 @@ const (
 	Other        Type = "Other"   // Other
 	Unknown           = "Unknown"
 	OtherUnknown      = string(Other + Unknown)
+)
+
+type Sex rune
+
+const (
+	Male   Sex = 'M'
+	Female     = 'F'
 )
 
 type Person struct {

--- a/utils_go18.go
+++ b/utils_go18.go
@@ -1,0 +1,30 @@
+//go:build go1.18
+// +build go1.18
+
+package swag
+
+import (
+	"reflect"
+	"unicode/utf8"
+)
+
+// AppendUtf8Rune appends the UTF-8 encoding of r to the end of p and
+// returns the extended buffer. If the rune is out of range,
+// it appends the encoding of RuneError.
+func AppendUtf8Rune(p []byte, r rune) []byte {
+	return utf8.AppendRune(p, r)
+}
+
+type CanIntegerValue struct {
+	reflect.Value
+}
+
+// CanInt reports whether Uint can be used without panicking.
+func (v CanIntegerValue) CanInt() bool {
+	return v.Value.CanInt()
+}
+
+// CanUint reports whether Uint can be used without panicking.
+func (v CanIntegerValue) CanUint() bool {
+	return v.Value.CanUint()
+}

--- a/utils_go18.go
+++ b/utils_go18.go
@@ -15,6 +15,7 @@ func AppendUtf8Rune(p []byte, r rune) []byte {
 	return utf8.AppendRune(p, r)
 }
 
+// CanIntegerValue a wrapper of reflect.Value
 type CanIntegerValue struct {
 	reflect.Value
 }

--- a/utils_other.go
+++ b/utils_other.go
@@ -3,6 +3,11 @@
 
 package swag
 
+import (
+	"reflect"
+	"unicode/utf8"
+)
+
 // AppendUtf8Rune appends the UTF-8 encoding of r to the end of p and
 // returns the extended buffer. If the rune is out of range,
 // it appends the encoding of RuneError.

--- a/utils_other.go
+++ b/utils_other.go
@@ -16,6 +16,7 @@ func AppendUtf8Rune(p []byte, r rune) []byte {
 	return p
 }
 
+// CanIntegerValue a wrapper of reflect.Value
 type CanIntegerValue struct {
 	reflect.Value
 }

--- a/utils_other.go
+++ b/utils_other.go
@@ -1,0 +1,41 @@
+//go:build !go1.18
+// +build !go1.18
+
+package swag
+
+// AppendUtf8Rune appends the UTF-8 encoding of r to the end of p and
+// returns the extended buffer. If the rune is out of range,
+// it appends the encoding of RuneError.
+func AppendUtf8Rune(p []byte, r rune) []byte {
+	length := utf8.RuneLen(rune(r))
+	if length > 0 {
+		utf8Slice := make([]byte, length)
+		utf8.EncodeRune(utf8Slice, rune(r))
+		p = append(p, utf8Slice...)
+	}
+	return p
+}
+
+type CanIntegerValue struct {
+	reflect.Value
+}
+
+// CanInt reports whether Uint can be used without panicking.
+func (v CanIntegerValue) CanInt() bool {
+	switch v.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return true
+	default:
+		return false
+	}
+}
+
+// CanUint reports whether Uint can be used without panicking.
+func (v CanIntegerValue) CanUint() bool {
+	switch v.Kind() {
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return true
+	default:
+		return false
+	}
+}

